### PR TITLE
New index checker

### PR DESCRIPTION
**Pull request by sijad**
I hope I haven't messed something up

### CODE REVIEW

**Comment by sijad** [comment]
I'm not a very good GO programmer but I think it's way cleaner
I'm working to add check if plugins sorted alphabetically

**Comment by bucaran** [comment]
This is really cool @sijad! Would you take some time summarize the changes?

* Added this.
* Removed this in favor of ...
* Fix that.

Etc.

**Comment by sijad** [comment]
sure
* removing compiled file in Favor of `go get` and Github releases
* organisation files
* add Colors to outputs
* cleanup codes
* cleanup outputs
* use a separated struct and method to verify each line to make code more effectives
* keep main function clean for feature improvements
* ...

**Comment by sijad** [comment]
also some test cases should be added, I'll work on that too

**Comment by bucaran** [comment]
@sijad I see, thanks. 

@jethrokuan is all yours.

**Comment by jethrokuan** [comment]
@sijad I really like your changes! I was just playing around initially. Good job :+1: 

I like how you've abstracted things out. I was thinking of updating the code to something similar to what you have right now.

Thing is, we need the binary file for the `.travis.yml` in the fisherman index repo. Of course, we could have travis fetch the checker source and build it, but that's another level of indirection.

In any case, feel free to hack away! You're doing a better job than I ever did :smile:

**Comment by jethrokuan** [comment]
@sijad one more issue:

I noticed you changed the error reporting style. It's not something I'm very against, but I think @bucaran  requested then to make it look like [TAP output](https://testanything.org/).

I'd also like to request that you try and avoid giving generic error prints like "Error occurred". Delete all the bad code I wrote! :stuck_out_tongue:

**Comment by sijad** [comment]
@jethrokuan
honesty I didn't  work with travis so much, can't we just run `go get github.com/fisherman/index-check` in `.travis.yml` script?
I'll change back outputs to be like TAP and sure try to cleanup more 😄

**Comment by bucaran** [comment]
@sijad So to summarize the main changes:

* Back to TAP (so no need for color)
* Make sure travis still works.

**Comment by sijad** [comment]
@jethrokuan can you please do me a Favor and try this as travis file, and make sure it works?

```
sudo: false
script:
  - go get github.com/fisherman/index-check; index-check ./index --no-color

```

@bucaran I did add colors output in case other users want test index in their local, an option `--no-color` also added to disable color output (this option doesn't work right now, I'll fix that)

**Comment by bucaran** [comment]
@sijad Color is fine, but if we revert to TAP, then we don't need color, but I guess a --no-color option is fine too. So, knock yourself out :smile:

**Comment by jethrokuan** [comment]
Looks like `.travis.yml` would have to look like this:

```
language: go

go:
  - tip
install:
  - go get github.com/fisherman/index-check
```

**Comment by sijad** [comment]
Updated
* colors removed
* output is now TAP again

File: indexLine.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+type indexLine struct {
+  no int
+  code indexCode
+  text string
+}
+
+func (l *indexLine) verify() error {
+	// Checks if fisher index is correctly formatted: 5 lines + 1 blank
+	if l.text == "" && l.no % 6 != 5 { // should change l.no
+		log.Fatalln("Fisher index is incorrectly formatted, at line:", l.no)
```
Comment (jethrokuan) [inline indexLine.go]: log.Fatalf("line %d: index is incorrectly formatted", l.no)

File: indexLine.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	}
+
+	switch l.code {
+	case indexName:
+		// Check if Name is Unique
+		if indexNames[l.text] {
+			return &indexLineError{"Name is not Unique", l.text, l.no}
```
Comment (jethrokuan) [inline indexLine.go]: return &indexLineError{"Name is not unique", l.text, l.no}

File: indexLine.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+			// return &indexLineError{"Index is not Sorted", l.text, l.no}
+		// }
+
+		// lastName = l.text
+
+		indexNames[l.text] = true
+	case indexLink:
+		// Check if URL is Unique
+		if indexUrls[l.text] {
+			return &indexLineError{"Url is not Unique", l.text, l.no}
```
Comment (jethrokuan) [inline indexLine.go]: return &indexLineError{"Url is not unique", l.text, l.no}

File: indexLine.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+		if len(invalids) != 0 {
+			b.WriteString("Invalid tags: " + strings.Join(invalids, ", "))
+		}
+
+		if len(duplicates) != 0 {
+			if b.Len() != 0 {
+				b.WriteString("    \n   ")
+			}
+			b.WriteString("Duplicate Tags: " + strings.Join(duplicates, ", "))
```
Comment (jethrokuan) [inline indexLine.go]: b.WriteString("Duplicate tags: " + strings.Join(duplicates, ", "))

File: indexLine.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+		if b.Len() != 0 {
+			return &indexLineError{b.String(), l.text, l.no}
+		}
+	case indexAuthor:
+	}
+
+	return nil
+}
+
+type indexLineError struct {
```
Comment (jethrokuan) [inline indexLine.go]: Could you move this on top? I think it's cleaner that way.

**Comment by jethrokuan** [comment]
@sijad I added some comments. After the cleanup, I think we're good to merge. I cloned your repo and confirmed that they all work. Tests can come in a separate PR (:

Thanks for your contribution!

**Comment by sijad** [comment]
@jethrokuan thanks for review

File: indexLine.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+	switch l.code {
+	case indexName:
+		// Check if Name is Unique
+		if indexNames[l.text] {
+			return &indexLineError{"Name is not unique", l.text, l.no}
+		}
+
+		// if lastName > l.text {
+			// return &indexLineError{"Index is not Sorted", l.text, l.no}
```
Comment (sijad) [inline indexLine.go]: shouldn't package names sorted alphabetically? current index seems is not sorted

File: indexLine.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+
+		indexUrls[l.text] = true
+
+		if urlMatcher.MatchString(l.text) != true {
+			return &indexLineError{"Url in not valid", l.text, l.no}
+		}
+	case indexDesc:
+		// Check Description
+		// if _, ok := regexp.MatchString(`^[\w\s]+$`, l.text); ok != nil {
+			// return &indexLineError{"Description is not valid", l.text, l.no}
```
Comment (sijad) [inline indexLine.go]: shouldn't description have some qualities like length, char set, ...?

File: indexLine.go (hunk 6 above)
Comment (jethrokuan) [inline indexLine.go]: It's not a hard requirement. At the moment the last item is not, and the index works fine. Alphabetical sorting is a nice thing to have, but not a must.

File: indexLine.go (hunk 7 above)
Comment (jethrokuan) [inline indexLine.go]: I'm not sure what measure we should impose here, but a check whether it's blank would be good. In fact, if it were blank the program would quit fatally now. 

It probably should have some qualities, but there are none that we have imposed as of yet, so it's not right for the checker to yell at them when we  didn't warn them whatsoever (:

File: indexLine.go (hunk 6 above)
Comment (jethrokuan) [inline indexLine.go]: @bucaran I realized not being in alphabetical order affects `fisher search`

File: indexLine.go (hunk 7 above)
Comment (sijad) [inline indexLine.go]: checking empty lines done in [line#55](https://github.com/sijad/index-check/blob/new-index-checker/indexLine.go#L55)

Comment (jethrokuan) [inline indexLine.go]: That's what I meant. So I think we can leave the description part out for now. thus far descriptions have been short and meaningful.

Maybe limit the number of characters, to 60 or 80 perhaps?

Comment (sijad) [inline indexLine.go]: 60 or 80 seems good, but I'm not sure, should we limit for tags number too?

Comment (jethrokuan) [inline indexLine.go]: Sure, I think about 6 is right.

@bucaran should have a better idea on how the index is supposed to look/function.

File: indexLine.go (hunk 6 above)
Comment (bucaran) [inline indexLine.go]: @jethrokuan Search does not sort the index when curling the index into the file. What about adding the quicksort snippet I wrote for `submit`?

File: indexLine.go (hunk 7 above)
Comment (bucaran) [inline indexLine.go]: There are no technical limits, but the online search looks better with fewer tags, 3 or 4, is best.

Comment (sijad) [inline indexLine.go]: what about maximum length of each tag? is 15 char good?

**Comment by sijad** [comment]
Script now checks sorted packages and also makes sure maximum number tags is 4

File: indexLine.go (hunk 7 above)
Comment (jethrokuan) [inline indexLine.go]: Sounds alright to me.

**Comment by bucaran** [comment]
@sijad @jethrokuan Are we ready to merge guys?

**Comment by sijad** [comment]
Ok
some few changes
* Cleaned up a little
* Now it check for longs tags
I guess it's ready

**Comment by bucaran** [comment]
@sijad Thanks! You are a collaborator now so feel free to merge it if you and @jethrokuan think is ready! 

Good job you guys!

**Comment by jethrokuan** [comment]
Looks good, gonna merge it. Thanks @sijad for all the effort!

**Pull request merged**